src/Middleware/JsonSchema.php

```diff
@@ -0,0 +1,98 @@
+<?php
+
+namespace Psr7Middlewares\Middleware;
+
+use JsonSchema\Validator;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+
+class JsonSchema
+{
+    /** @var string[] */
+    private $schemas;
+
+    /**
+     * JsonSchema constructor.
+     * @param string[] $schemas [uri => file]
+     */
+    public function __construct(array $schemas)
+    {
+        $this->schemas = $schemas;
+    }
+
+    /**
+     * Execute the middleware.
+     *
+     * @param ServerRequestInterface $request
+     * @param ResponseInterface      $response
+     * @param callable               $next
+     *
+     * @return ResponseInterface
+     */
+    public function __invoke(ServerRequestInterface $request, ResponseInterface $response, callable $next)
+    {
+        if ($schema = $this->getSchema($request)) {
+            $validator = new Validator();
+            $decoded = $this->decodeBody($request);
+            if (json_last_error() !== JSON_ERROR_NONE) {
+                return $response->withStatus(422, 'Malformed JSON: ' . json_last_error_msg());
+            }
+
+            $validator->check($decoded, (object) [
+                '$ref' => $this->normalizeSchema($schema),
+            ]);
+
+            if (!$validator->isValid()) {
+                $response->getBody()->write(stripslashes(json_encode($validator->getErrors())));
+
+                return $response->withStatus(422, 'Unprocessable Entity')
+                    ->withHeader('Content-Type', 'application/json');
+            }
+        }
+
+        return $next ? $next($request, $response) : $response;
+    }
+
+    /**
+     * @param ServerRequestInterface $request
+     * @return string|bool
+     */
+    private function getSchema(ServerRequestInterface $request)
+    {
+        foreach ($this->schemas as $pattern => $file) {
+            $path = $request->getUri()->getPath();
+            if (stripos($path, $pattern) === 0) {
+                return $file;
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * @param string
+     * @return string
+     */
+    private function normalizeSchema($schema)
+    {
+        if (parse_url($schema, PHP_URL_SCHEME)) {
+            // The schema file already has a scheme, e.g. `file://` or `vfs://`.
+            return $schema;
+        }
+
+        return 'file://' . $schema;
+    }
+
+    /**
+     * \Psr\Http\Message\MessageInterface::getParsedBody returns a multi-dimensional array, but
+     * \JsonSchema\Validator::check requires a recursive object.
+     *
+     *
+     * @param ServerRequestInterface $request
+     * @return object
+     */
+    private function decodeBody(ServerRequestInterface $request)
+    {
+        return json_decode($request->getBody(), $assoc = false);
+    }
```

Comment:
I understand the reason of this, but I prefer to delegate the body parsing to other middleware instead. There's a Payload middleware available that, by default, parses the body to array, but we can provide a config setting to change this behaviour.